base/inference.jl

```diff
@@ -1629,43 +1543,119 @@ function typeinf_edge(method::Method, atypes::ANY, sparams::SimpleVector, needtr
         end
     end
     typeinf_loop(frame)
-    ccall(:jl_typeinf_end, Void, ())
-    return (frame.src, widenconst(frame.bestguess), frame.inferred)
+    return frame
 end
 
-function typeinf_edge(method::Method, atypes::ANY, sparams::SimpleVector, caller)
-    return typeinf_edge(method, atypes, sparams, false, true, true, caller)
-end
-function typeinf(method::Method, atypes::ANY, sparams::SimpleVector, needtree::Bool=false)
-    return typeinf_edge(method, atypes, sparams, needtree, true, true, nothing)
-end
-# compute an inferred (optionally optimized) AST without global effects (i.e. updating the cache)
-function typeinf_uncached(method::Method, atypes::ANY, sparams::ANY; optimize::Bool=true)
-    return typeinf_edge(method, atypes, sparams, true, optimize, false, nothing)
+# compute (and cache) an inferred AST and return the current best estimate of the result type
+function typeinf_edge(method::Method, atypes::ANY, sparams::SimpleVector, caller::InferenceState)
+    code = code_for_method(method, atypes, sparams)
+    code === nothing && return Any
+    code = code::MethodInstance
+    if isdefined(code, :inferred)
+        inf = code.inferred
+        if isa(inf, CodeInfo)
+            if (inf::CodeInfo).inferred
+                return code.rettype
+            end
+        else
+            return code.rettype
```

Comment:
agreed 3. I'm not sure why Github doesn't hide the comments anymore after the code is changed.